test/e2e/receive_test.go

```diff
@@ -12,97 +13,123 @@ import (
-				Add(receiver(1, defaultPromRemoteWriteConfig(nodeExporterHTTP(1), remoteWriteEndpoint(1)), 3, remoteWriteEndpoint(1), remoteWriteEndpoint(2), remoteWriteEndpoint(3))).
-				Add(receiver(2, defaultPromConfig("no-remote-write", 2), 3, remoteWriteEndpoint(1), remoteWriteEndpoint(2), remoteWriteEndpoint(3))).
-				Add(receiver(3, defaultPromConfig("no-remote-write", 3), 3, remoteWriteEndpoint(1), remoteWriteEndpoint(2), remoteWriteEndpoint(3)))
-	receiveReplicationMetrics = []model.Metric{
-		{
-			"__name__": "up",
-			"instance": model.LabelValue(nodeExporterHTTP(1)),
-			"job":      "node",
-			"receive":  "true",
-			"replica":  model.LabelValue("1"),
-		},
-		{
-			"__name__": "up",
-			"instance": model.LabelValue(nodeExporterHTTP(1)),
-			"job":      "node",
-			"receive":  "true",
-			"replica":  model.LabelValue("2"),
-		},
-		{
-			"__name__": "up",
-			"instance": model.LabelValue(nodeExporterHTTP(1)),
-			"job":      "node",
-			"receive":  "true",
-			"replica":  model.LabelValue("3"),
-		},
-	}
-)
-
 type receiveTestConfig struct {
-	testConfig
+	name    string
+	cmds    []scheduler
 	metrics []model.Metric
+
+	queryAddress address
+}
+
+func defaultPromRemoteWriteConfig(nodeExporterHTTP, remoteWriteEndpoint string) string {
+	return fmt.Sprintf(`
+scrape_configs:
+- job_name: 'node'
+  static_configs:
+  - targets: ['%s']
+ remote_write:
+ - url: "%s"
+`, nodeExporterHTTP, remoteWriteEndpoint)
 }
 
 func TestReceive(t *testing.T) {
-	for _, tt := range []receiveTestConfig{
-		{
-			testConfig{
-				"hashring",
-				receiveHashringSuite,
-			},
-			receiveHashringMetrics,
-		},
-		{
-			testConfig{
-				"replication",
-				receiveReplicationSuite,
-			},
-			receiveReplicationMetrics,
-		},
-	} {
+	var testCases []receiveTestConfig
+
+	a := newLocalAddresser()
+	{
+		// The hashring suite creates three receivers, each with a Prometheus
+		// remote-writing data to it. However, due to the hashing of the labels,
+		// the time series from the Prometheus is forwarded to a different
+		// receiver in the hashring than the one handling the request.
+		// The querier queries all the receivers and the test verifies
+		// the time series are forwarded to the correct receive node.
+
+		receiveHTTP1, receiveHTTP2, receiveHTTP3 := a.New(), a.New(), a.New()
+
+		r1 := receiver(receiveHTTP1, a.New(), a.New(), 1, remoteWriteEndpoint(receiveHTTP1), remoteWriteEndpoint(receiveHTTP2), remoteWriteEndpoint(receiveHTTP3))
+		r2 := receiver(receiveHTTP2, a.New(), a.New(), 1, remoteWriteEndpoint(receiveHTTP1), remoteWriteEndpoint(receiveHTTP2), remoteWriteEndpoint(receiveHTTP3))
+		r3 := receiver(receiveHTTP3, a.New(), a.New(), 1, remoteWriteEndpoint(receiveHTTP1), remoteWriteEndpoint(receiveHTTP2), remoteWriteEndpoint(receiveHTTP3))
+
+		prom1 := prometheus(a.New(), defaultPromRemoteWriteConfig(nodeExporterHTTP(1), remoteWriteEndpoint(r1.HTTP)))
```

Comment:
Do we need to still have both `addRemoteWriteToConfig(defaultPromConfig(...))` as well as `defaultPromRemoteWriteConfig`?

Reply:
Nope, removed.